Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.mm

```diff
@@ -853,39 +841,51 @@ void getSupportedTypes(HashSet<String>& types) const final
 {
     // Decompression sessions do not support encrypted content; force layer
     // creation.
-    if (m_mediaSourcePrivate && (m_mediaSourcePrivate->cdmInstance() || m_mediaSourcePrivate->needsVideoLayer()))
+    if (!canUseDecompressionSession())
         return true;
-    auto player = m_player.get();
+    RefPtr player = m_player.get();
     return player && player->renderingCanBeAccelerated();
 }
 
 void MediaPlayerPrivateMediaSourceAVFObjC::setPresentationSize(const IntSize& newSize)
 {
     if (!layerOrVideoRenderer() && !newSize.isEmpty())
-        updateDisplayLayerAndDecompressionSession();
+        updateDisplayLayer();
 }
 
 void MediaPlayerPrivateMediaSourceAVFObjC::setVideoLayerSizeFenced(const FloatSize& newSize, WTF::MachSendRight&&)
 {
     if (!layerOrVideoRenderer() && !newSize.isEmpty())
-        updateDisplayLayerAndDecompressionSession();
+        updateDisplayLayer();
 }
 
 void MediaPlayerPrivateMediaSourceAVFObjC::acceleratedRenderingStateChanged()
 {
-    updateDisplayLayerAndDecompressionSession();
+    RefPtr renderer = layerOrVideoRenderer();
+    if (willUseDecompressionSessionIfNeeded()) {
+        if (renderer) {
+            if (m_mediaSourcePrivate)
+                m_mediaSourcePrivate->videoRendererWillReconfigure(*renderer);
+            renderer->setPrefersDecompressionSession(true);
+            if (m_mediaSourcePrivate)
+            m_mediaSourcePrivate->videoRendererDidReconfigure(*renderer);
+            return;
+        }
+    }
+    if (renderer && willUseDecompressionSessionIfNeeded())
+        return; // With a decompression session we can continue using the existing VideoMediaSampleRenderer. No need to re-create one.
```

Comment:
This is kind of confusing. You should never hit the second `if()` case because it will have returned early inside the `Igf (renderer)` clause.